### PR TITLE
eth/tracers: fix the issue prestate missing existing contract state

### PR DESCRIPTION
**Pull request by jsvisa**
See more from https://github.com/ethereum/go-ethereum/pull/25422#discussion_r978789901

### CODE REVIEW

File: eth/tracers/native/prestate.go [hunk 1]
```diff
@@ -116,8 +120,12 @@ func (t *prestateTracer) CaptureStart(env *vm.EVM, from common.Address, to commo
 
 // CaptureEnd is called after the call finishes to finalize the tracing.
 func (t *prestateTracer) CaptureEnd(output []byte, gasUsed uint64, _ time.Duration, err error) {
-	if t.create && !t.config.DiffMode {
-		// Exclude created contract.
+	if t.config.DiffMode {
+		return
+	}
+
+	if t.create && !t.pre[t.to].isExists() {
```
Comment (holiman) [inline eth/tracers/native/prestate.go:127]: Shouldn't you do a similar `if s := t.pre[a]; s != nil {` ?

Comment (jsvisa) [reply eth/tracers/native/prestate.go:127]: sorry ：）

Comment (jsvisa) [reply eth/tracers/native/prestate.go:127]: Seems we don't need to care about the `to_address` nullable, in `diffMode`, the created address may be cleared if nothing is modified. But in prestate-only mode, we don't need to care about that.


But for consistency, let's check it.

File: eth/tracers/native/prestate.go [hunk 2]
```diff
@@ -45,6 +45,10 @@ type account struct {
 	Storage map[common.Hash]common.Hash `json:"storage,omitempty"`
 }
 
+func (a *account) isExists() bool {
```
Comment (s1na) [inline eth/tracers/native/prestate.go:48]: Nitpick: to stay consistent rename to `empty` as in stateObject and change accordingly:

https://github.com/ethereum/go-ethereum/blob/d86fe26f67d8ad17d8b66da9a8908101fb4c11cb/core/state/state_object.go#L98-L100

Also note the `balance.Sign() == 0` check. Balance < 0 should not happen but it still makes account non-empty.

**Comment by holiman** [comment]
The term empty is a consensus-defined term, important for the empty-touch-clear functionality. The usage here is not related, it's simply "did this account seem to exist prior to the create?"

**Comment by holiman** [comment]
Further, storage plays no part in the consensus-definition of empty. So either stay consistent and use the same definition and term, or use different checks and a different term

**Comment by jsvisa** [comment]
agree +1, rename back to `exists`

**Review by holiman — APPROVED** [review]
LGTM (if you change back empty to something else)

**Review by s1na — APPROVED** [review]
LGTM. I updated the `exists` semantics based on the name.

**Pull request merged**